src/query/storages/fuse/src/io/read/block/parquet/mod.rs

```diff
@@ -0,0 +1,147 @@
+use databend_common_expression::DataBlock;
+use databend_common_expression::TableDataType;
+use databend_common_expression::TableSchema;
+use databend_common_expression::Value;
+use databend_storages_common_cache::CacheAccessor;
+use databend_storages_common_cache::TableDataCacheKey;
+use databend_storages_common_cache_manager::CacheManager;
+use databend_storages_common_table_meta::meta::ColumnMeta;
+use databend_storages_common_table_meta::meta::Compression;
+
+mod adapter;
+mod deserialize;
+
+use crate::io::read::block::block_reader_merge_io::DataItem;
+use crate::io::read::block::parquet::deserialize::deserialize_column_chunks;
+use crate::io::BlockReader;
+impl BlockReader {
+    pub(crate) fn column_chunks_to_data_block_1(
+        &self,
+        num_rows: usize,
+        column_metas: &HashMap<ColumnId, ColumnMeta>,
+        column_chunks: HashMap<ColumnId, DataItem>,
+        compression: &Compression,
+        block_path: &str,
+    ) -> databend_common_exception::Result<DataBlock> {
+        if column_chunks.is_empty() {
+            return self.build_default_values_block(num_rows);
+        }
+        let record_batch = deserialize_column_chunks(
+            &self.original_schema,
+            num_rows,
+            &column_chunks,
+            compression,
+        )?;
+        let mut columns = Vec::with_capacity(self.projected_schema.fields.len());
+        let name_paths = column_name_paths(&self.projection, &self.original_schema);
+        for (i, field) in self.projected_schema.fields.iter().enumerate() {
+            let data_type = field.data_type().into();
+            let leaf_column_ids = field.leaf_column_ids();
+            if leaf_column_ids.len() > 1
+                && leaf_column_ids
+                    .iter()
+                    .any(|id| matches!(column_chunks.get(id), Some(DataItem::ColumnArray(_))))
+            {
+                return Err(ErrorCode::StorageOther(
+                    "unexpected nested field: nested leaf field hits cached",
+                ));
+            }
+
+            let value = match column_chunks.get(&field.column_id) {
+                Some(DataItem::RawData(data)) => {
+                    let arrow_array = column_by_name(&record_batch, &name_paths[i]);
+                    let arrow2_array: Box<dyn databend_common_arrow::arrow::array::Array> =
+                        arrow_array.into();
+                    if self.put_cache && leaf_column_ids.len() == 1 {
+                        if let Some(cache) = CacheManager::instance().get_table_data_array_cache() {
+                            let meta = column_metas.get(&field.column_id).unwrap();
+                            let (offset, len) = meta.offset_length();
+                            let key =
+                                TableDataCacheKey::new(block_path, field.column_id, offset, len);
+                            cache.put(key.into(), Arc::new((arrow2_array.clone(), data.len())))
```

Comment:
seems the size of data (array) being cached is not kept correctly, if the field being processed is nested.

since the data item (DataItem::RawData) we got from `column_chunks` is always data of a leave column.

besides, the following test failed, if array cache is enabled:


~~~
mysql> create table t1(c tuple(c1 int not null, c2 int null) null);
Query OK, 0 rows affected (0.12 sec)

mysql> insert into t1 values((1,null));
Query OK, 1 row affected (0.11 sec)

mysql> select c.1 from t1;
+------+
| c.1  |
+------+
|    1 |
+------+
1 row in set (0.05 sec)
Read 1 rows, 4.00 B in 0.010 sec., 95.94 rows/sec., 383.78 B/sec.

mysql> select c from t1;
ERROR 1105 (HY000): StorageOther. Code: 4000, Text = unexpected nested field: nested leaf field hits cached.

mysql> select * from system.configs where name = 'table_data_deserialized_data_bytes';
+-------+------------------------------------+----------+-------------+
| group | name                               | value    | description |
+-------+------------------------------------+----------+-------------+
| cache | table_data_deserialized_data_bytes | 10240000 |             |
+-------+------------------------------------+----------+-------------+
1 row in set (0.01 sec)
Read 168 rows, 9.98 KiB in 0.006 sec., 28.68 thousand rows/sec., 1.66 MiB/sec.
~~~

Reply:
Fixed:
```
MySQL [(none)]> select * from system.configs where name = 'table_data_deserialized_data_bytes';
+-------+------------------------------------+--------+-------------+
| group | name                               | value  | description |
+-------+------------------------------------+--------+-------------+
| cache | table_data_deserialized_data_bytes | 102400 |             |
+-------+------------------------------------+--------+-------------+
1 row in set (0.030 sec)

MySQL [(none)]> create table t1(c tuple(c1 int not null, c2 int null) null);
Query OK, 0 rows affected (0.044 sec)

MySQL [(none)]> insert into t1 values((1,null));
Query OK, 1 row affected (0.074 sec)

MySQL [(none)]> select c.1 from t1;
+------+
| c.1  |
+------+
|    1 |
+------+
1 row in set (0.049 sec)

MySQL [(none)]> select c from t1;
+----------+
| c        |
+----------+
| (1,NULL) |
+----------+
1 row in set (0.054 sec)

MySQL [(none)]> select c from t1;
+----------+
| c        |
+----------+
| (1,NULL) |
+----------+
1 row in set (0.055 sec)

MySQL [(none)]> select c.1 from t1;
+------+
| c.1  |
+------+
|    1 |
+------+
1 row in set (0.038 sec)
```